File: analytics/zone_analysis.py
"""
zone_analysis.py
-----------------
Batch PySpark job: computes zone-wise parking utilization and average
parking duration, and ranks zones from most to least utilized.

Run:
    spark-submit analytics/zone_analysis.py
"""

from pathlib import Path

from pyspark.sql import SparkSession
from pyspark.sql.functions import avg, col

BASE_DIR = Path(__file__).resolve().parent.parent
RAW_CSV = str(BASE_DIR / "data" / "raw" / "parking_events.csv")
OUTPUT_CSV = str(BASE_DIR / "data" / "processed" / "zone_utilization")


def main():
    spark = SparkSession.builder.appName("ParkingZoneAnalysis").getOrCreate()
    spark.sparkContext.setLogLevel("WARN")

    df = (
        spark.read
        .option("header", True)
        .option("inferSchema", True)
        .csv(RAW_CSV)
    )

    utilization = df.withColumn("utilization", (col("occupancy") / col("capacity")) * 100)

    zone_utilization = (
        utilization.groupBy("zone_id")
        .agg(avg("utilization").alias("average_utilization"))
        .orderBy(col("average_utilization").desc())
    )

    print("Zone-wise average utilization:")
    zone_utilization.show(50, truncate=False)

    avg_duration = (
        df.filter(col("event_type") == "EXIT")
        .groupBy("zone_id")
        .agg(avg("parking_duration").alias("average_parking_duration_min"))
        .orderBy(col("average_parking_duration_min").desc())
    )

    print("Average parking duration per zone (minutes):")
    avg_duration.show(50, truncate=False)

    zone_utilization.coalesce(1).write.mode("overwrite").option("header", True).csv(OUTPUT_CSV)
    print(f"Zone utilization written to {OUTPUT_CSV}")


if __name__ == "__main__":
    main()

File: r-analytics/zone_analysis.R
# zone_analysis.R
# Computes zone-wise parking utilization and average parking duration.
# R equivalent of analytics/zone_analysis.py.
#
# Run (from the project root):
#   Rscript r-analytics/zone_analysis.R

suppressMessages(library(dplyr))

RAW_CSV <- "data/raw/parking_events.csv"
OUTPUT_CSV <- "r-analytics/output/zone_utilization_r.csv"

if (!file.exists(RAW_CSV)) {
  stop(sprintf("%s not found. Run `python scripts/generate_dataset.py` first.", RAW_CSV))
}

df <- read.csv(RAW_CSV, stringsAsFactors = FALSE)
df$utilization <- df$occupancy / df$capacity * 100

zone_utilization <- df %>%
  group_by(zone_id) %>%
  summarise(average_utilization = mean(utilization), .groups = "drop") %>%
  arrange(desc(average_utilization))

cat("Zone-wise average utilization:\n")
print(zone_utilization, n = 50)

avg_duration <- df %>%
  filter(event_type == "EXIT") %>%
  group_by(zone_id) %>%
  summarise(average_parking_duration_min = mean(parking_duration), .groups = "drop") %>%
  arrange(desc(average_parking_duration_min))

cat("\nAverage parking duration per zone (minutes):\n")
print(avg_duration, n = 50)

dir.create(dirname(OUTPUT_CSV), showWarnings = FALSE, recursive = TRUE)
write.csv(zone_utilization, OUTPUT_CSV, row.names = FALSE)
cat(sprintf("\nSaved to %s\n", OUTPUT_CSV))
